### PR TITLE
ci: generate headless bundles using esbuild

### PR DESCRIPTION
**Pull request by samisayegh**
Build time (my machine): ~1 second (esbuild only)
Build time (Jenkins): 109 seconds (esbuild-rollup hybrid) vs. 234 seconds (rollup only)
Esbuild minified bundle size: ~3% larger

### Bundle size comparison (minified)
Percentage comparisons are versus rollup.

|Use Case | Rollup (kb) | Esbuild (kb) | Diff (%) | Terser minification | Diff (%)
|---|---|---|---|---|---
|search|243|260| 7 | 255 | 5
|recommendation|176|182| 3 | 178 | 1
|product recommendation|177|183| 3
|product listing|193|199 | 3
|case-assist|179|181 | 1

Terser provides smaller bundles at the cost of additional time and configuration complexity due to using two tools (not included in this PR). Rollup still outperforms. I only ran the terser benchmark for two use-cases.

**Summary**
Salesforce needs a UMD bundle, so using esbuild OOTB is not possible. It might be [possible to wrap](https://github.com/evanw/esbuild/issues/507#issuecomment-727690399) the `iife` bundle to make it a UMD, but there might some unknown pitfalls. I prefer to explore this in a separate PR.

The hybrid solution slows the production build (rollup umd generation takes ~1min locally), but is still significantly faster. One benefit though is a smaller UMD bundle size, which is helpful since UMD bundles are typically consumed fully (no treeshaking).
Dev builds are very fast because they are powered purely by esbuild.

### CODE REVIEW

File: packages/headless/package.json [hunk 1]
```diff
@@ -71,6 +71,8 @@
     "@rollup/plugin-replace": "2.4.2",
     "@types/jest": "27.0.3",
     "concurrently": "6.4.0",
+    "esbuild": "^0.14.1",
```
Comment (samisayegh) [inline packages/headless/package.json:74]: Esbuild build is still `0.x`. Additional information around [production readiness](https://esbuild.github.io/faq/#production-readiness) available here.

**Comment by ThibodeauJF** [comment]
The time it takes to build is 1 second vs 4mins (approx. info could be relevant in the description) ? That's amazing!

File: packages/headless/package.json (hunk 1 above)
Comment (btaillon) [reply packages/headless/package.json:74]: I'm fine with that, in this case.

**Comment by btaillon** [comment]
Here are my `gtime -v` (`gnu-time`) results:
| esbuild | rollup |
| --- | --- |
| image | image |

Memory usage is down from ~3.5gbytes in rollup to ~0.5gbytes in esbuild.

I was checking, because I ran out of memory in github actions for the Headless build.

**Review by ThibodeauJF — APPROVED** [review]

**Review by btaillon — APPROVED** [review]

**Comment by btaillon** [comment]
Apparently that ESBuild doesn't do type checking? If that is so, you should probably add a type checking step to the CI. I'll migrate it over to github actions.

**Comment by samisayegh** [comment]
> Apparently that ESBuild doesn't do type checking?
Yep, that is correct. As discussed in the scrum though, we are still typechecking as part of the step that creates the declaration files.

**Pull request merged**